Crust/Mapper/MappingOperator.swift

```diff
@@ -246,91 +263,170 @@ private func mapFromJson<T, U: Mapping>(_ json: JSONValue, toField field: inout
+                                     using: mapping,
+                                     fieldContains: fieldContains,
+                                     context: map.context)
+        }
+        else if let baseJSON = baseJSON {
+            newObjects = try generateNewValues(fromJsonArray: baseJSON,
+                                     with: updatePolicy,
+                                     using: mapping,
+                                     fieldContains: fieldContains,
+                                     context: map.context)
+        }
+        else {
+            let userInfo = [ NSLocalizedFailureReasonErrorKey : "JSON at key path \(map.key) does not exist to map from" ]
+            throw NSError(domain: CrustMappingDomain, code: 0, userInfo: userInfo)
+        }
+        
+        return (newObjects, map.context)
 }
 
-private func mapFromJson<T, U: Mapping, V: RangeReplaceableCollection>(_ json: JSONValue, toField field: inout V, mapping: U, context: MappingContext, allowDuplicates: Bool) throws where U.MappedObject == T, V.Iterator.Element == T, T: Equatable {
+private func generateNewValues<T, U: Mapping, S: Sequence>(
+    fromJsonArray json: JSONValue,
+    with updatePolicy: CollectionUpdatePolicy<S>,
+    using mapping: U,
+    fieldContains: (T) -> Bool,
+    context: MappingContext)
+    throws -> [T]
+    where U.MappedObject == T, T: Equatable, U.SequenceKind.Iterator.Element == U.MappedObject {
     
-    if case .array(let xs) = json {
+        guard case .array(let xs) = json else {
+            let userInfo = [ NSLocalizedFailureReasonErrorKey : "Trying to map json of type \(type(of: json)) to \(U.SequenceKind.self)<\(T.self)>" ]
+            throw NSError(domain: CrustMappingDomain, code: -1, userInfo: userInfo)
+        }
+        
         let mapper = Mapper<U>()
-        var results = [T]()
+        
+        var newObjects = [T]()
+        
+        let isUnique = { (obj: T, newObjects: [T], fieldContains: (T) -> Bool) -> Bool in
+            let newObjectsContainsObj = newObjects.contains(obj)
+            
+            switch updatePolicy.insert {
+            case .replace(_):
+                return !newObjectsContainsObj
+            case .append:
+                return !(newObjectsContainsObj || fieldContains(obj))
+            }
+        }
+        
         for x in xs {
-            if !allowDuplicates {
-                if let obj = try mapping.getExistingInstance(json: x) {
-                    if results.contains(obj) {
-                        continue
-                    }
+            let obj = try mapper.map(from: x, using: mapping, parentContext: context)
+            
+            switch updatePolicy.unique {
+            case false:
+                newObjects.append(obj)
+                
+            case true:
+                if isUnique(obj, newObjects, fieldContains) {
+                    newObjects.append(obj)
                 }
             }
```

Comment:
Iterate, map, and add to an array of new objects. We prevent mutating the original field being mapped to prevent copies on that field. Once we have all the newly mapped data we then either append or simply assign straight to the field.